File: crates/netstatus/src/captive_portal.rs
use serde::Serialize;
use std::time::Duration;
use ureq::Agent;

const CAPTIVE_PORTAL_URL: &str = "http://captive.apple.com/hotspot-detect.html";
const CAPTIVE_PORTAL_TIMEOUT: Duration = Duration::from_secs(10);
const EXPECTED_BODY_MARKER: &str = "Success";

/// Whether something between this machine and the internet is
/// intercepting plain HTTP traffic to serve a login page (a captive
/// portal), determined by fetching Apple's own hotspot-detect endpoint —
/// the same check macOS itself uses.
#[derive(Debug, Clone, Copy, PartialEq, Serialize)]
pub enum CaptivePortalStatus {
    Clear,
    Detected,
    Unknown,
}

/// Apple's endpoint returns a 200 with "Success" in the body when nothing
/// intercepted the request. Matching on the marker rather than the full HTML
/// byte-for-byte avoids false `Detected` results from whitespace/formatting
/// drift Apple has introduced across OS versions — the same tolerance
/// Apple's own client uses. Anything else — a rewritten body, a redirect,
/// any other status — means a portal (or something else) is altering the
/// response.
fn classify_response(status: u16, body: &str) -> CaptivePortalStatus {
    if status == 200 && body.contains(EXPECTED_BODY_MARKER) {
        CaptivePortalStatus::Clear
    } else {
        CaptivePortalStatus::Detected
    }
}

/// Fetches Apple's captive-portal-check endpoint over plain HTTP — not
/// HTTPS, since a captive portal intercepts/rewrites unencrypted HTTP and
/// this probe needs that interception to be visible. A transport-level
/// failure (timeout, DNS failure, connection refused) means the check
/// itself didn't complete, not that a portal was found, so it maps to
/// `Unknown` rather than `Detected`.
pub fn check_captive_portal() -> CaptivePortalStatus {
    let config = Agent::config_builder()
        .timeout_global(Some(CAPTIVE_PORTAL_TIMEOUT))
        .build();
    let agent: Agent = config.into();

    let Ok(mut response) = agent.get(CAPTIVE_PORTAL_URL).call() else {
        return CaptivePortalStatus::Unknown;
    };
    let status = response.status().as_u16();
    let Ok(body) = response.body_mut().read_to_string() else {
        return CaptivePortalStatus::Unknown;
    };
    classify_response(status, &body)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn exact_success_body_is_clear() {
        assert_eq!(
            classify_response(
                200,
                "<HTML><HEAD><TITLE>Success</TITLE></HEAD><BODY>Success</BODY></HTML>"
            ),
            CaptivePortalStatus::Clear
        );
    }

    #[test]
    fn altered_body_is_detected() {
        assert_eq!(
            classify_response(
                200,
                "<HTML><HEAD><TITLE>Portal</TITLE></HEAD><BODY>Login required</BODY></HTML>"
            ),
            CaptivePortalStatus::Detected
        );
    }

    #[test]
    fn bare_success_without_html_wrapper_is_clear() {
        assert_eq!(
            classify_response(200, "Success"),
            CaptivePortalStatus::Clear
        );
    }

    #[test]
    fn non_200_status_is_detected() {
        assert_eq!(
            classify_response(302, "Success"),
            CaptivePortalStatus::Detected
        );
    }
}

File: apps/NetCheck/Sources/NetStatus/CaptivePortal.swift
import Foundation

extension Probe {
    private static let captivePortalURLString = "http://captive.apple.com/hotspot-detect.html"
    private static let captivePortalTimeout: TimeInterval = 10

    /// Whether something between this machine and the internet is
    /// intercepting plain HTTP — determined by fetching Apple's own
    /// hotspot-detect endpoint over HTTP (not HTTPS: the interception has
    /// to be visible) and checking for its `Success` marker. A transport
    /// failure is inconclusive (`.unknown`), not evidence of a portal.
    public static func checkCaptivePortal() async -> CaptivePortalStatus {
        await checkCaptivePortal(fetch: liveCaptivePortalFetch)
    }

    static func checkCaptivePortal(
        fetch: @Sendable (URL) async throws -> (Data, URLResponse)
    ) async -> CaptivePortalStatus {
        guard let url = URL(string: captivePortalURLString) else { return .unknown }
        do {
            let (data, response) = try await fetch(url)
            guard let http = response as? HTTPURLResponse else { return .unknown }
            // Malformed bytes in a portal's HTML should degrade to replacement
            // characters, not abort classification — the lossy initializer is
            // deliberate here.
            // swiftlint:disable:next optional_data_string_conversion
            let body = String(decoding: data, as: UTF8.self)
            return classifyCaptivePortalResponse(status: http.statusCode, body: body)
        } catch {
            return .unknown
        }
    }

    private static func liveCaptivePortalFetch(_ url: URL) async throws -> (Data, URLResponse) {
        var request = URLRequest(url: url)
        request.timeoutInterval = captivePortalTimeout
        request.cachePolicy = .reloadIgnoringLocalAndRemoteCacheData

        let session = URLSession(configuration: .ephemeral)
        defer { session.finishTasksAndInvalidate() }

        return try await session.data(for: request, delegate: RedirectRefuser())
    }

    /// A 200 carrying `Success` means nothing intercepted the request.
    /// Anything else — rewritten body, redirect, other status — is a
    /// portal (or something else altering the response).
    static func classifyCaptivePortalResponse(status: Int, body: String) -> CaptivePortalStatus {
        status == 200 && body.contains("Success") ? .clear : .detected
    }
}

/// Refuses HTTP redirects so a captive portal's `302 -> login page` is
/// seen as the 302 it is, rather than followed to a page that might
/// coincidentally 200.
private final class RedirectRefuser: NSObject, URLSessionTaskDelegate, Sendable {
    func urlSession(
        _: URLSession,
        task _: URLSessionTask,
        willPerformHTTPRedirection _: HTTPURLResponse,
        newRequest _: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        completionHandler(nil)
    }
}
